### PR TITLE
Finger Paint App

### PR DESCRIPTION
**Pull request by CraftsmanMG**
Add fingerPaint App code and resources.
**This current version still needs to be updated to work outside of the hifi repo.**

In order for the fingerPaint App to run successfully, there's a dependency with engine code that is currently under a merge with hifi:master  [PR](https://github.com/highfidelity/hifi/pull/10702).

### CODE REVIEW

**Comment by themelissabrown** [comment]
Engine side PR has been merged. We know there is an issue with the sphere not always rezzing when changing hands but would like to merge this anyway and apply a bugfix for the QA fail on https://github.com/highfidelity/hifi/pull/10702

**Comment by Nex-Pro** [comment]
I feel this is a **PASS** as the issues seen are minor and IMO not blockers.

Used this link to test along with five other users.

https://rawgit.com/CraftsmanMG/hifi-content/0fbfdb9689479263f97c387bc009a6c6196922ad/Marketplace/fingerPaint/fingerPaint.js


All features were tested and seemed to work fine for all users. We had a large group working together. We even tried co-ordinating our strokes to closely follow the leader and this worked really well. Everyone  enjoyed using finger paint.
Everyone would be happier with a larger number of undos.

Had a request for a hifi logo brush.

We could identify each other's work when declaring what we would draw.

Avatar scaling with hand controllers is blocked when finger paint is active.

**Undo:** Occasionally when undoing a delete operation the user can see the restored stroke but others do not see it. We had this occur with three observers.

**Hue Animation:** Occasionally strokes do not animate. The failed strokes are observed by all.

**Mouse mode:** Does not always work despite reloading scripts.

**Continous line:** And possibly any line fails when the avatar is moving at speed.

**Brushes:** Brush masks are slow to load (User perception) Seems to take a couple of seconds.

**Deleting:** Sometimes with very long and undulating strokes it can be difficult to find a spot on the line which will trigger the delete.

There are a few minor gripes.

It is easy to unintentionally place a stroke when grabbing the tablet. Ie. you can draw when moving the tablet with the dominant hand.

Sometimes random polyline vert positions are created if the stroke intersects the tablet.

**Performance:**
User experience even with large numbers of strokes in view ( hundreds ) is good with no one reporting noticeable frame rate reductions.

There is no obvious way to delete ( apart from undo ) when painting in desktop mode.

There would probably ( almost certainly ) be issues with attempting to select large numbers of the strokes in the ent list but that is not due to this PR.

Tie a ribbon round this and send it on it's way.

***Any previous feedback I've given regarding the quality of the strokes and other features still applies.*** 

hifi-snap-by-piper-peppercorn-on-2017-11-10_00-30-22

File: Marketplace/fingerPaint/content/animatedBrushes/animatedBrush.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
+//This file is just used for the superclass
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedBrush.js]: Please add copyright headers

File: Marketplace/fingerPaint/content/animatedBrushes/animatedBrush.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+function AnimatedBrushClass() {}
+
+/**
+ * Called on every frame draw after the user stops painting
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedBrush.js]: Follow code style guidelines for commenting

File: Marketplace/fingerPaint/content/animatedBrushes/animatedBrush.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ * 
+ * @abstract
+ * @param deltaSeconds, the number of seconds past since the last frame was rendered
+ * @param entityID: the id of the polyline being drawn
+ * @param fingerOverlayID: the id of the overlay that shows over the finger when using fingerpaint
+ */
+AnimatedBrushClass.prototype.onUpdate = function(deltaSeconds, entityID) {
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedBrush.js]: Indentation should follow coding guidelines

File: Marketplace/fingerPaint/content/animatedBrushes/animatedBrush.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+/**
+ * This function updates the user data so in the next frame the animation gets the previous values.\
+ *
+ * @param entityID: the id of the polyline being animated
+ * @param animatedBrushObject: the animation object (should be a subclass of animatedBrush)
+ */
+AnimatedBrushClass.prototype.updateUserData = function(entityID, animatedBrushObject) {
+    var prevUserData = Entities.getEntityProperties(entityID).userData;
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedBrush.js]: Only request the properties required from Entities.getEntityProperties

File: Marketplace/fingerPaint/content/animatedBrushes/animatedBrushScript.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
+(function() {
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedBrushScript.js]: Copyright header

File: Marketplace/fingerPaint/content/animatedBrushes/animatedBrushScript.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
+(function() {    
+    Script.include("animatedBrushesList.js");
+    var UPDATE_TIME = 33; //run at aproximatelly 30fps
+    var MIN_PLAY_DISTANCE = 6; //Minimum distance from player to entity in order to play animation
+    var self = this;
+    this.preload = function(entityID) {
+        print("After adding script 2 : " + JSON.stringify(Entities.getEntityProperties(entityID).color));
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedBrushScript.js]: Remove print statements or wrap in a debug catch. Only request necessary entity properties with getEntityProperties()

File: Marketplace/fingerPaint/content/animatedBrushes/animatedBrushScript.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        self.intervalID = Script.setInterval(function() {
+            if (MyAvatar.sessionUUID != Entities.getEntityProperties(entityID).lastEditedBy) {
+                Script.clearInterval(self.intervalID);
+                return;
+            }
+
+            if (Vec3.withinEpsilon(MyAvatar.position, Entities.getEntityProperties(entityID).position, MIN_PLAY_DISTANCE)) {
+                var userData = Entities.getEntityProperties(entityID).userData;
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedBrushScript.js]: Only request necessary entity properties

File: Marketplace/fingerPaint/content/animatedBrushes/animatedBrushScript.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+                if (userData) {
+                    var userDataObject = JSON.parse(userData);
+                    var animationObject = userDataObject.animations;
+                    var newAnimationObject = null;
+                    if (!userDataObject.timeFromLastAnimation) {
+                        userDataObject.timeFromLastAnimation = Date.now();
+                    }
+                    Object.keys(animationObject).forEach(function(animationName) {
+                        newAnimationObject = animationObject[animationName];
+                        //print("Proto 0001: " + JSON.stringify(newAnimationObject));
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedBrushScript.js]: Remove commented out code

File: Marketplace/fingerPaint/content/animatedBrushes/animatedBrushesList.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
+Script.include("animatedHueBrush.js");
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedBrushesList.js]: Copyright header

File: Marketplace/fingerPaint/content/animatedBrushes/animatedBrushesList.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        case "animatedHueBrush":
+            return new AnimatedHueBrush(settings, entityID);
+        case "animatedRotationBrush":
+            return new AnimatedRotationBrush(settings, entityID);
+        case "animatedTranslationBrush":
+            return new AnimatedTranslationBrush(settings, entityID);
+        default:
+            throw new Error("Invalid Brush Name. Could not instantiate " + animatedBrushName);
+    }
+}
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedBrushesList.js]: EOF newline

File: Marketplace/fingerPaint/content/animatedBrushes/animatedHueBrush.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
+//Superclass
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedHueBrush.js]: Copyright header

File: Marketplace/fingerPaint/content/animatedBrushes/animatedHueBrush.js [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+Script.include("animatedBrush.js");
+Script.include("../js/ColorUtils.js");
+
+function AnimatedHueBrushClass(settings, entityID) {
+    AnimatedBrush.call(this);
+    this.hsvColor = rgb2hsv(Entities.getEntityProperties(entityID).color);//   {hue: 0, saturation: 1.0, value: 1.0};
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedHueBrush.js]: Only request required properties

File: Marketplace/fingerPaint/content/animatedBrushes/animatedHueBrush.js [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    this.animatedColor = {red: 0, green: 0, blue: 0};
+}
+
+AnimatedHueBrushClass.prototype.ANIMATED_BRUSH_TIME = 10; //inteval in milliseconds to update the brush width;
+AnimatedHueBrushClass.prototype.ANIMATED_BRUSH_INCREMENT = 0.5; //linear increment of brush size;
+AnimatedHueBrushClass.prototype.NAME = "animatedHueBrush"; 
+
+AnimatedHueBrushClass.prototype.onUpdate = function(deltaSeconds, entityID) {
+	print( "My animation brushes are updating ");
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedHueBrush.js]: Remove print statements or wrap in debug

File: Marketplace/fingerPaint/content/animatedBrushes/animatedHueBrush.js [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    this.hsvColor.hue = this.hsvColor.hue >= 360 ? 0 : this.hsvColor.hue; //restart hue cycle
+    this.animatedColor = hsv2rgb(this.hsvColor);
+    Entities.editEntity(entityID, { color : this.animatedColor});
+    this.parent.updateUserData(entityID, this);
+}
+
+AnimatedHueBrushClass.prototype.constructor = AnimatedHueBrushClass;
+AnimatedHueBrushClass.prototype.parent = AnimatedBrush.prototype;
+
+AnimatedHueBrush = AnimatedHueBrushClass;
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedHueBrush.js]: EOF newline

File: Marketplace/fingerPaint/content/animatedBrushes/animatedRotationBrush.js [hunk 15]
```diff
@@ -0,0 +1,25 @@
+//Superclass
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedRotationBrush.js]: Copyright header

File: Marketplace/fingerPaint/content/animatedBrushes/animatedRotationBrush.js [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+
+AnimatedRotationBrushClass.prototype.onUpdate = function(deltaSeconds, entityID) {
+    this.angle = this.angle + ((deltaSeconds * this.ANIMATED_BRUSH_INCREMENT)/this.ANIMATED_BRUSH_TIME);
+    this.angle = this.angle >= 360 ? 0 : this.angle; //restart hue cycle
+    var rotation = Vec3.multiply(this.angle, this.activeAxis);
+    Entities.editEntity(entityID, {rotation : Quat.fromPitchYawRollDegrees(rotation.x, rotation.y, rotation.z)});
+    this.parent.updateUserData(entityID, this);
+}
+
+AnimatedRotationBrush = AnimatedRotationBrushClass;
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedRotationBrush.js]: EOF newline

File: Marketplace/fingerPaint/content/animatedBrushes/animatedRotationBrush.js [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+AnimatedRotationBrushClass.prototype.constructor = AnimatedRotationBrushClass;
+AnimatedRotationBrushClass.prototype.parent = AnimatedBrush.prototype;
+
+AnimatedRotationBrushClass.prototype.ANIMATED_BRUSH_TIME = 100; //inteval in milliseconds to update the entity rotation;
+AnimatedRotationBrushClass.prototype.ANIMATED_BRUSH_INCREMENT = 5; //linear increment of brush size;
+AnimatedRotationBrushClass.prototype.NAME = "animatedRotationBrush"; //linear increment of brush size;
+
+AnimatedRotationBrushClass.prototype.onUpdate = function(deltaSeconds, entityID) {
+    this.angle = this.angle + ((deltaSeconds * this.ANIMATED_BRUSH_INCREMENT)/this.ANIMATED_BRUSH_TIME);
+    this.angle = this.angle >= 360 ? 0 : this.angle; //restart hue cycle
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedRotationBrush.js]: values should be named constants

File: Marketplace/fingerPaint/content/animatedBrushes/animatedTranslationBrush.js [hunk 18]
```diff
@@ -0,0 +1,41 @@
+//Superclass
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedTranslationBrush.js]: Copyright header

File: Marketplace/fingerPaint/content/animatedBrushes/animatedTranslationBrush.js [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+    if (Vec3.distance(nextPosition, this.startingPosition) > this.MAX_TRANSLATION) {
+        this.translation = 0;
+        nextPosition = this.startingPosition;
+    }
+    Entities.editEntity(entityID, {position : nextPosition});
+    this.parent.updateUserData(entityID, this);
+}
+
+AnimatedTranslationBrush = AnimatedTranslationBrushClass;
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedTranslationBrush.js]: EOF newline

File: Marketplace/fingerPaint/content/animatedBrushes/animatedTranslationBrush.js [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+AnimatedTranslationBrushClass.prototype.constructor = AnimatedTranslationBrushClass;
+AnimatedTranslationBrushClass.prototype.parent = AnimatedBrush.prototype;
+
+AnimatedTranslationBrushClass.prototype.ANIMATED_BRUSH_TIME = 10; //inteval in milliseconds to update the brush width;
+AnimatedTranslationBrushClass.prototype.ANIMATED_BRUSH_INCREMENT = 0.005; //linear increment of brush size;
+AnimatedTranslationBrushClass.prototype.MAX_TRANSLATION = 2;
+AnimatedTranslationBrushClass.prototype.NAME = "animatedTranslationBrush"; //linear increment of brush size;
+
+AnimatedTranslationBrushClass.prototype.onUpdate = function(deltaSeconds, entityID) {
+    var currentPosition = Entities.getEntityProperties(entityID).position;
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/animatedBrushes/animatedTranslationBrush.js]: Only request required entity properties

File: Marketplace/fingerPaint/content/js/ColorUtils.js [hunk 21]
```diff
@@ -0,0 +1,76 @@
+//converts hsv color into rgb color space, expects hsv with the following ranges
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/js/ColorUtils.js]: Copyright header

File: Marketplace/fingerPaint/content/js/ColorUtils.js [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+//H(0, 359), S(0, 1), V(0, 1) to R(0, 255), G(0, 255), B(0, 255)
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/js/ColorUtils.js]: Please name constants for values in this in this file

File: Marketplace/fingerPaint/content/js/ColorUtils.js [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+    if (cMax == 0) {
+        hsvColor.saturation = 0;
+    } else {
+        hsvColor.saturation = (deltaC/cMax);
+    }
+
+    hsvColor.value = (cMax);
+
+    return hsvColor;
+}
```
Comment (misslivirose) [inline Marketplace/fingerPaint/content/js/ColorUtils.js]: EOF newline

File: Marketplace/fingerPaint/fingerPaint.js [hunk 24]
```diff
@@ -0,0 +1,1143 @@
...
+            _strokeNormals,
+            _strokeColors,
+            _strokeWidths,
+            _timeOfLastPoint,
+            _isContinuousLine = _savedSettings.currentIsContinuous,
+            _lastPosition = null,
+            _shouldKeepDrawing = false,
+            _texture = CONTENT_PATH + "/" + _savedSettings.currentTexture.brushName,
+            _dynamicEffects = _savedSettings.currentDynamicBrushes;
+            //'https://upload.wikimedia.org/wikipedia/commons/thumb/9/93/Caris_Tessellation.svg/1024px-Caris_Tessellation.svg.png', // Daantje
```
Comment (misslivirose) [inline Marketplace/fingerPaint/fingerPaint.js]: remove commented out properties

File: Marketplace/fingerPaint/fingerPaint.js [hunk 25]
```diff
@@ -0,0 +1,1143 @@
...
+            if (_isBrushColored) {
+                properties.color = {red: 255, green: 255, blue: 255};
+                return;
+            }
+
+            var isAnyDynamicEffectEnabled = false;
+            if ("dynamicHue" in _dynamicEffects && _dynamicEffects.dynamicHue) {
+                isAnyDynamicEffectEnabled = true;
+                var hueIncrement = 359.0 / MAX_POINTS_PER_LINE;
+                _dynamicColor.hue = calculateValueInRange(_dynamicColor.hue, 0, 359, hueIncrement);
```
Comment (misslivirose) [inline Marketplace/fingerPaint/fingerPaint.js]: Name constants

File: Marketplace/fingerPaint/fingerPaint.js [hunk 26]
```diff
@@ -0,0 +1,1143 @@
...
+            } 
+
+            if ("dynamicSaturation" in _dynamicEffects && _dynamicEffects.dynamicSaturation) {
+                isAnyDynamicEffectEnabled = true;
+                _dynamicColor.saturation = _dynamicColor.saturation == 0.5 ? 1.0 : 0.5;
```
Comment (misslivirose) [inline Marketplace/fingerPaint/fingerPaint.js]: Name constants

File: Marketplace/fingerPaint/fingerPaint.js [hunk 27]
```diff
@@ -0,0 +1,1143 @@
...
+                //saturation along the full line
```
Comment (misslivirose) [inline Marketplace/fingerPaint/fingerPaint.js]: Remove commented out code

File: Marketplace/fingerPaint/fingerPaint.js [hunk 28]
```diff
@@ -0,0 +1,1143 @@
+//
```
Comment (misslivirose) [inline Marketplace/fingerPaint/fingerPaint.js]: Name constant values in this file and remove commented out code

File: Marketplace/fingerPaint/fingerPaint.js [hunk 29]
```diff
@@ -0,0 +1,1143 @@
...
+                opositeHandPosition = MyAvatar.getJointPosition(handName === "left" ? "RightHandMiddle1" : "LeftHandMiddle1");
+                
+                if (triggerValue < TRIGGER_START_WIDTH_RAMP) {
+                    lineWidth = MIN_LINE_WIDTH;
+                } else {
+                    lineWidth = MIN_LINE_WIDTH
+                        + (triggerValue - TRIGGER_START_WIDTH_RAMP) / TRIGGER_RAMP_WIDTH * RAMP_LINE_WIDTH;
+                }
+                
+                if ((handName === "left" && _isLeftHandDominant) || (handName === "right" && !_isLeftHandDominant)){
```
Comment (misslivirose) [inline Marketplace/fingerPaint/fingerPaint.js]: space after closing ) before {

File: Marketplace/fingerPaint/fingerPaint.js [hunk 30]
```diff
@@ -0,0 +1,1143 @@
...
+    function updateHandAnimations(){
+        if (_leftBrush == null) {
+            return;
+        }
+        var ANIM_URL = (_isLeftHandDominant? LEFT_ANIM_URL: RIGHT_ANIM_URL );
+        var ANIM_OPEN = (_isLeftHandDominant? LEFT_ANIM_URL_OPEN: RIGHT_ANIM_URL_OPEN );
+        var handLiteral = (_isLeftHandDominant? "left": "right" );
+
+        //Clear previous hand animation override
+        restoreAllHandAnimations();
```
Comment (misslivirose) [inline Marketplace/fingerPaint/fingerPaint.js]: Name animation constants

File: Marketplace/fingerPaint/fingerPaint.js [hunk 31]
```diff
@@ -0,0 +1,1143 @@
...
+        MyAvatar.overrideRoleAnimation(handLiteral + "IndexPointAndThumbRaiseClosed", ANIM_URL, 30, false, 19, 20);
+        
+        //turn off lasers and other interactions
+        Messages.sendLocalMessage("Hifi-Hand-Disabler", "none");
+        Messages.sendLocalMessage("Hifi-Hand-Disabler", handLiteral);
+        
+        //update ink Source
+
+        var strokeColor = _leftBrush.getStrokeColor();
+        var strokeWidth = _leftBrush.getStrokeWidth()*0.06;
```
Comment (misslivirose) [inline Marketplace/fingerPaint/fingerPaint.js]: Name this number

File: Marketplace/fingerPaint/fingerPaint.js [hunk 32]
```diff
@@ -0,0 +1,1143 @@
...
+        updateHandAnimations();
+        
+        // Messages channels for enabling/disabling other scripts' functions.
+        Messages.subscribe(HIFI_POINT_INDEX_MESSAGE_CHANNEL);
+        Messages.subscribe(HIFI_GRAB_DISABLE_MESSAGE_CHANNEL);
+        Messages.subscribe(HIFI_POINTER_DISABLE_MESSAGE_CHANNEL);
+    }
+    
+    function enableProcessing() {
+        print("INFO: Star Processing");
```
Comment (misslivirose) [inline Marketplace/fingerPaint/fingerPaint.js]: Should this be "Start"?

File: Marketplace/fingerPaint/fingerPaint.js [hunk 33]
```diff
@@ -0,0 +1,1143 @@
...
+            //Restores and clears hand animations
+            restoreAllHandAnimations();
+            
+            //clears Overlay sphere
+            Overlays.deleteOverlay(_inkSourceOverlay);
+            _inkSourceOverlay = null;
+        }
+    }
+
+    //Load last fingerpaint settings
```
Comment (misslivirose) [inline Marketplace/fingerPaint/fingerPaint.js]: Name variables

File: Marketplace/fingerPaint/fingerPaint.js [hunk 34]
```diff
@@ -0,0 +1,1143 @@
...
+                    event.brushName =  Script.resolvePath(event.brushName);
+                    _leftBrush.setTexture(event.brushName);
+                    _rightBrush.setTexture(event.brushName);
+                }
+                break;
+
+            case "changeLineWidth":
+                if (_leftBrush != null && _rightBrush != null) {
+                    Settings.setValue("currentStrokeWidth", event.brushWidth);
+                    var dim = event.brushWidth * 2 + 0.1;
```
Comment (misslivirose) [inline Marketplace/fingerPaint/fingerPaint.js]: variables should be named constants

File: Marketplace/fingerPaint/fingerPaint.js [hunk 35]
```diff
@@ -0,0 +1,1143 @@
...
+
+            default:
+                break;
+        }
+    }    
+
+    function addAnimationToBrush(entityID) {
+        Object.keys(AnimatedBrushesInfo).forEach(function(animationName) {
+            if (AnimatedBrushesInfo[animationName].isEnabled) {
+                var prevUserData = Entities.getEntityProperties(entityID).userData;
```
Comment (misslivirose) [inline Marketplace/fingerPaint/fingerPaint.js]: Only request necessary properties

File: Marketplace/fingerPaint/fingerPaint.js [hunk 36]
```diff
@@ -0,0 +1,1143 @@
...
+
+        tablet.screenChanged.disconnect(onTabletScreenChanged);
+        tablet.tabletShownChanged.disconnect(onTabletShownChanged);
+        button.clicked.disconnect(onButtonClicked);
+        tablet.removeButton(button);
+    }
+   
+    function getFingerPosition(x, y) {
+        var pickRay = Camera.computePickRay(x, y);
+        return Vec3.sum(pickRay.origin, Vec3.multiply(pickRay.direction, 5));
```
Comment (misslivirose) [inline Marketplace/fingerPaint/fingerPaint.js]: 5 as named constant

File: Marketplace/fingerPaint/hifi/js/eventBridgeLoader.js [hunk 37]
```diff
@@ -0,0 +1,19 @@
+
```
Comment (misslivirose) [inline Marketplace/fingerPaint/hifi/js/eventBridgeLoader.js]: Copyright header

File: Marketplace/fingerPaint/hifi/js/keyboardControl.js [hunk 38]
```diff
@@ -0,0 +1,67 @@
...
+    function doLowerKeyboard() {
+        EventBridge.emitWebEvent("_LOWER_KEYBOARD");
+        lowerTimer = null;
+        isRaised = false;
+    }
+
+    function lowerKeyboard() {
+        // Delay lowering keyboard a little in case immediately raise it again.
+        if (lowerTimer === null) {
+            lowerTimer = setTimeout(doLowerKeyboard, 20);
```
Comment (misslivirose) [inline Marketplace/fingerPaint/hifi/js/keyboardControl.js]: name constant

**Comment by misslivirose** [comment]
Closing as these changes are encompassed in #132 with the required structure for app certification

**Pull request closed without merging**